anomaly_detector/core/pipeline.py

```diff
@@ -0,0 +1,121 @@
+        else:
+            raise TypeError()
+
+    def execute_steps(self, tracer=None):
+        """Execute steps one by one."""
+        for step in self.steps:
+            if tracer is None:
+                step.execute()
+            else:
+                step.execute_with_tracing(tracer)
+            self.count += 1
+
+    def __len__(self):
+        """Number of steps in queue."""
+        return len(self.steps)
+
+    def clear(self):
+        """Reset the values."""
+        self.steps = []
+        self.count = 0
+
+
+class PipelineCatalog(object):
+    """Model Catalog for selecting which MLModel To Use for Anomaly Detection."""
+
+    def __init__(self, config, feedback_strategy, job):
+        """Pipeline initialization logic."""
+        self.config = config
+        self.feedback_strategy = feedback_strategy
+        if job in self._class_method_choices:
+            self.job = job
+        else:
+            raise ValueError("Unsupported job used {}".format(job))
+
+    @classmethod
+    def create_sompy_modeladapter(cls, config, feedback_strategy):
+        """Setup sompy model adapter which provides functionality required to train SOMPY Model with W2V encoding."""
+        if feedback_strategy is None:
+            feedback_strategy = FeedbackStrategy(config=config)
+        storage_adapter = SomStorageAdapter(config, feedback_strategy)
+        model_adapter = SomModelAdapter(storage_adapter)
+        return model_adapter
+
+    @classmethod
+    def _sompy_train_job(cls, config, feedback_strategy):
+        """Perform Training and inference of SOMPY Model."""
+        pipeline = Pipeline()
+        model_adapter = cls.create_sompy_modeladapter(config, feedback_strategy)
+        train = SomTrainJob(node_map=config.SOMPY_NODE_MAP, model_adapter=model_adapter)
+        pipeline.add_steps(train)
+        return pipeline
+
+    @classmethod
+    def _sompy_train_infer_job(cls, config, feedback_strategy):
+        """Perform Training and inference of SOMPY Model."""
+        pipeline = Pipeline()
+        model_adapter = cls.create_sompy_modeladapter(config, feedback_strategy)
+        pipeline.add_steps(SomTrainJob(node_map=config.SOMPY_NODE_MAP, model_adapter=model_adapter))
+        pipeline.add_steps(SomInferenceJob(model_adapter=model_adapter))
+        return pipeline
+
+    @classmethod
+    def _sompy_infer_job(cls, config, feedback_strategy):
+        """Perform inference of SOMPY Model."""
+        pipeline = Pipeline()
+        model_adapter = cls.create_sompy_modeladapter(config, feedback_strategy)
+        train = SomTrainJob(node_map=config.SOMPY_NODE_MAP, model_adapter=model_adapter)
+        pipeline.add_steps(train)
+        return pipeline
+
+    _class_method_choices = {'sompy.train': _sompy_train_job,
+                             'sompy.inference': _sompy_infer_job,
+                             'sompy.train.inference': _sompy_train_infer_job}
+
+    def get_pipeline(self):
+        """Pipeline Api api."""
```

Comment:
What does this method do?

Reply:
This returns the pipeline that is used to train the model. We will be different algorithms soon so this would need to be the extension point.

Reply:
Can you add the explanation to the doc string then?

Reply:
Done.